### PR TITLE
Refactor AuthLogPlugin

### LINKED ISSUE #286: Auth Log format update on rfc5424
1. Preamble 
   
   The basic format for syslog messages in the "/var/log/auth.log" file follows the standard syslog format. 
   
   ```
   <Timestamp> <Hostname> <Service>: <Message>
   ```
   
   The following pattern applies specifically to the auth.log file.
   
   ```
   <Timestamp> <Hostname> <Service>[<process_id>]: <Message>
   ```
   
   See: https://datatracker.ietf.org/doc/html/rfc5424#section-6
   
   1. Test data
      
      ```log
      Mar 29 10:43:01 ip-10-77-20-248 sshd[1193]: Accepted password for test_user from REDACTED IP port 52942 ssh2
      Mar 27 13:06:56 ip-10-77-20-248 sshd[1291]: Server listening on REDACTED IP port 22.
      Mar 27 13:06:56 ip-10-77-20-248 systemd-logind[1118]: Watching system buttons on /dev/input/event0 (Power Button)
      Mar 27 13:06:56 ip-10-77-20-248 systemd-logind[1118]: Watching system buttons on /dev/input/event1 (Sleep Button)
      Mar 27 13:06:56 ip-10-77-20-248 systemd-logind[1118]: New seat seat0.
      Mar 27 13:08:09 ip-10-77-20-248 sshd[1361]: Accepted publickey for test_user from REDACTED IP port 12345 ssh2: RSA SHA256:123456789asdfghjklöertzuio
      Mar 27 13:08:09 ip-10-77-20-248 sshd[1361]: pam_unix(sshd:session): session opened for user test_user by (uid=0)
      Mar 27 13:08:09 ip-10-77-20-248 systemd: pam_unix(systemd-user:session): session opened for user test_user by (uid=0)
      Mar 27 13:10:08 ip-10-77-20-248 sudo:   ubuntu : TTY=pts/0 ; PWD=/home/test_user ; USER=root ; COMMAND=/usr/bin/apt-key add -
      Mar 27 13:17:01 ip-10-77-20-248 CRON[2623]: pam_unix(cron:session): session closed for user root
      Mar 27 13:23:11 ip-10-77-20-248 sudo: pam_unix(sudo:session): session opened for user root by test_user(uid=0)
      ```
      
      
1. Problem with the current config
   
   The config of the regex pattern is declared in:
   
   https://github.com/fox-it/dissect.target/blob/main/dissect/target/plugins/os/unix/log/auth.py#L18
   
   But if acquire or target runs. The jsonl export deliver a pattern/object with: 
   
   - hostname
   - domain
   - ts
   - message 
   - source 
     
     The service and the process id is in this version part of the message.
     
     It would be better if the service (and process id) and the message will be seperated.
     
     Example for this line:
     
     ```
     Mar 29 10:43:01 ip-10-77-20-248 sshd[1193]: Accepted password for test_user from REDACTED IP port 52942 ssh2
     ```
     
     Declared is this standard in: https://datatracker.ietf.org/doc/html/rfc3164#section-5.4.
     Ses also: https://www.loggly.com/ultimate-guide/linux-logging-basics/
     
     
1. Possible Solution
   
   Try this pattern to the auth.log file and you will get the service and the process seperated from the message.
   
   (
   w{3}
   s+
   d+
   s+
   d{2}:
   d{2}:
   d{2})
   s+(
   S+)
   s+([0-9A-Za-z
   -]**)?
   [?(
   w**)?
   ]?:
   s+(.+)
   
   devided by groups 
   
   To access the data use:
   
   ```
   _AUTH_LOG_REGEX = r"^(\w{3}\s+\d+\s+\d{2}:\d{2}:\d{2})\s+(\S+)\s+([0-9A-Za-z\-]**)?\[?(\w**)?\]?:\s+(.+)"
   
   RE_ALRE = re.compile(_AUTH_LOG_REGEX)
   ...
       auth_log_object = re.search(RE_ALRE, line)
       if auth_log_object:
         print(auth_log_object[1])
         print(auth_log_object[2])
         print(auth_log_object[3])
         print(auth_log_object[4])
   ```
   
1. Upgrade for the code
   
   Try this example and use named subgroups for the regex.
   
   ```python
   
   import re
   
   auth_log_content = """
   Mar 29 10:43:01 my_unix_host sshd[1193]: Accepted password for test_user from REDACTED IP port 52942 ssh2
   Mar 27 13:06:56 my_unix_host sshd[1291]: Server listening on REDACTED IP port 22.
   Mar 27 13:06:56 my_unix_host systemd-logind[1118]: Watching system buttons on /dev/input/event0 (Power Button)
   Mar 27 13:06:56 my_unix_host systemd-logind[1118]: Watching system buttons on /dev/input/event1 (Sleep Button)
   Mar 27 13:06:56 my_unix_host systemd-logind[1118]: New seat seat0.
   Mar 27 13:08:09 my_unix_host sshd[1361]: Accepted publickey for test_user from REDACTED IP port 12345 ssh2: RSA SHA256:123456789asdfghjklöertzuio
   Mar 27 13:08:09 my_unix_host sshd[1361]: pam_unix(sshd:session): session opened for user test_user by (uid=0)
   Mar 27 13:08:09 my_unix_host systemd: pam_unix(systemd-user:session): session opened for user test_user by (uid=0)
   Mar 27 13:10:08 my_unix_host sudo:   ubuntu : TTY=pts/0 ; PWD=/home/test_user ; USER=root ; COMMAND=/usr/bin/apt-key add -
   Mar 27 13:17:01 my_unix_host CRON[2623]: pam_unix(cron:session): session closed for user root
   Mar 27 13:23:11 my_unix_host sudo: pam_unix(sudo:session): session opened for user root by test_user(uid=0)
   """
   
   _TS_REGEX = r"^[A-Za-z]{3}\s\*[0-9]{1,2}\s[0-9]{1,2}:[0-9]{2}:[0-9]{2}"
   RE_TS = re.compile(_TS_REGEX)
   RE_TS_AND_HOSTNAME = re.compile(_TS_REGEX + r"\s\S+\s")
   
   
   
   for line in auth_log_content.split("\n"):
     ts_and_hostname = re.search(RE_TS_AND_HOSTNAME, line)
     if ts_and_hostname:
       message = line.replace(ts_and_hostname.group(0), "").strip()
       print(ts_and_hostname.group(0), message)
   
   
   _AUTH_LOG_REGEX = r"^(\w{3}\s+\d+\s+\d{2}:\d{2}:\d{2})\s+(\S+)\s+([0-9A-Za-z\-]**)?\[?(\w**)?\]?:\s+(.+)"
   
   RE_ALRE = re.compile(_AUTH_LOG_REGEX)
   
   for line in auth_log_content.split("\n"):
     
     auth_log_object = re.search(RE_ALRE, line)
     
     if auth_log_object:
       print()
       print(line)
       print(auth_log_object[1])
       print(auth_log_object[2])
       print(auth_log_object[3])
       print(auth_log_object[4])
       print(auth_log_object[5])
   
   
   
   _AUTH_LOG_NAMED_REGEX = r"^(?P<timestamp>\w{3}\s+\d+\s+\d{2}:\d{2}:\d{2})\s+(?P<hostname>\S+)\s+(?P<service>[0-9A-Za-z\-]**)?\[?(?P<processid>\w**)?\]?:\s+(?P<message>.+)"
   
   RE_ALNRE = re.compile(_AUTH_LOG_NAMED_REGEX)
   
   for line in auth_log_content.split("\n"):
     
     auth_log_object = re.match(RE_ALNRE, line)
     
     if auth_log_object:
       details = auth_log_object.groupdict()
       print(details)
   ```
   
   Output 
   
   ```python
   {'timestamp': 'Mar 29 10:43:01', 'hostname': 'my_unix_host', 'service': 'sshd', 'processid': '1193', 'message': 'Accepted password for test_user from REDACTED IP port 52942 ssh2'}
   {'timestamp': 'Mar 27 13:06:56', 'hostname': 'my_unix_host', 'service': 'sshd', 'processid': '1291', 'message': 'Server listening on REDACTED IP port 22.'}
   {'timestamp': 'Mar 27 13:06:56', 'hostname': 'my_unix_host', 'service': 'systemd-logind', 'processid': '1118', 'message': 'Watching system buttons on /dev/input/event0 (Power Button)'}
   {'timestamp': 'Mar 27 13:06:56', 'hostname': 'my_unix_host', 'service': 'systemd-logind', 'processid': '1118', 'message': 'Watching system buttons on /dev/input/event1 (Sleep Button)'}
   {'timestamp': 'Mar 27 13:06:56', 'hostname': 'my_unix_host', 'service': 'systemd-logind', 'processid': '1118', 'message': 'New seat seat0.'}
   {'timestamp': 'Mar 27 13:08:09', 'hostname': 'my_unix_host', 'service': 'sshd', 'processid': '1361', 'message': 'Accepted publickey for test_user from REDACTED IP port 12345 ssh2: RSA SHA256:123456789asdfghjklöertzuio'}
   {'timestamp': 'Mar 27 13:08:09', 'hostname': 'my_unix_host', 'service': 'sshd', 'processid': '1361', 'message': 'pam_unix(sshd:session): session opened for user test_user by (uid=0)'}
   {'timestamp': 'Mar 27 13:08:09', 'hostname': 'my_unix_host', 'service': 'systemd', 'processid': '', 'message': 'pam_unix(systemd-user:session): session opened for user test_user by (uid=0)'}
   {'timestamp': 'Mar 27 13:10:08', 'hostname': 'my_unix_host', 'service': 'sudo', 'processid': '', 'message': 'ubuntu : TTY=pts/0 ; PWD=/home/test_user ; USER=root ; COMMAND=/usr/bin/apt-key add -'}
   {'timestamp': 'Mar 27 13:17:01', 'hostname': 'my_unix_host', 'service': 'CRON', 'processid': '2623', 'message': 'pam_unix(cron:session): session closed for user root'}
   {'timestamp': 'Mar 27 13:23:11', 'hostname': 'my_unix_host', 'service': 'sudo', 'processid': '', 'message': 'pam_unix(sudo:session): session opened for user root by test_user(uid=0)'}
   ```
   
   
   So I prefer the named subgroups. 
   But this is maybe not necessary for the mapping in the yield part.

### PR DESCRIPTION
**Pull request by Poeloe**
Fixes #286

### CODE REVIEW

**Comment by Poeloe** [comment]
Small elaboration:

Went for an approach where each record will always contain:
- `ts`
- `source`
- `service`
- `pid`
- `message`
- `_target`

I then try to parse additional fields based on the service that created the log line (where `pam_unix` is an exception). 
This parsing is on a "best effort" basis. So if it fails or does not find any additional fields, the record would always still contain the above fields. Next to that, the additional fields are added dynamically to the record.

File: dissect/target/plugins/os/unix/log/auth.py [hunk 1]
```diff
@@ -1,39 +1,268 @@
...
-        ("path", "source"),
-    ],
-)
-
 _TS_REGEX = r"^[A-Za-z]{3}\s*[0-9]{1,2}\s[0-9]{1,2}:[0-9]{2}:[0-9]{2}"
 RE_TS = re.compile(_TS_REGEX)
 RE_TS_AND_HOSTNAME = re.compile(_TS_REGEX + r"\s\S+\s")
 
 
+class AuthLogServicesEnum(str, Enum):
```
[the remark this answers is not in the document]
Comment (Schamper) [reply dissect/target/plugins/os/unix/log/auth.py:20]: Can you also add a comment as to what these services are for? If I understand correctly, they are to provide enhanced records for specific log lines?

File: dissect/target/plugins/os/unix/log/auth.py [hunk 2]
```diff
@@ -1,39 +1,264 @@
...
+    cron = "CRON"
+    su = "su"
+    sudo = "sudo"
+    sshd = "sshd"
+    systemd = "systemd"
+    systemd_logind = "systemd-logind"
+    pkexec = "pkexec"
+
+
+class AuthLogRecordBuilder:
```
Comment (Schamper) [inline dissect/target/plugins/os/unix/log/auth.py:30]: I'm not really a fan of this "god class". I'd prefer smaller individual functions or classes for each "enhanced" parser.

File: tests/plugins/os/unix/log/test_auth.py [hunk 3]
```diff
@@ -96,7 +96,193 @@ def test_auth_plugin_year_rollover(target_unix, fs_unix: VirtualFilesystem):
...
+                "pid": "2623",
+                "action": "session closed",
+                "user": "root",
+            },
+            id="cron: pam_unix",
+        ),
+    ],
+)
+def test_auth_plugin_additional_fields(
+    target_unix, fs_unix: VirtualFilesystem, tmp_path: Path, message: str, results: dict[str, Union[str, int]]
```
Comment (Schamper) [inline tests/plugins/os/unix/log/test_auth.py:278]:
```suggestion
    target_unix, fs_unix: VirtualFilesystem, tmp_path: Path, message: str, results: dict[str, str | int]
```

File: tests/plugins/os/unix/log/test_auth.py [hunk 4]
```diff
@@ -96,7 +96,193 @@ def test_auth_plugin_year_rollover(target_unix, fs_unix: VirtualFilesystem):
...
+):
```
Comment (Schamper) [inline tests/plugins/os/unix/log/test_auth.py:279]:
```suggestion
) -> None:
```

**Review by Schamper — CHANGES_REQUESTED** [review]

**Review by Schamper — CHANGES_REQUESTED** [review]

File: dissect/target/plugins/os/unix/log/auth.py [hunk 5]
```diff
@@ -1,60 +1,362 @@
...
+
+    def build_record(self, ts: datetime, source: Path, line: str) -> TargetRecordDescriptor:
+        """Builds an ``AuthLog`` event record."""
+
+        record_fields = [
+            ("datetime", "ts"),
+            ("path", "source"),
+            ("string", "service"),
+            # PID should be string, since it can be "None"
+            ("string", "pid"),
```
Comment (Schamper) [inline dissect/target/plugins/os/unix/log/auth.py:253]: Can't `flow.record` fields be `None` too?

File: dissect/target/plugins/os/unix/log/auth.py [hunk 6]
```diff
@@ -1,60 +1,362 @@
...
-                message = line.replace(ts_and_hostname.group(0), "").strip()
-                yield AuthLogRecord(
-                    ts=ts,
-                    message=message,
-                    source=auth_file,
-                    _target=self.target,
-                )
+            if is_iso_fmt(auth_file):
+                iterable = iso_readlines(auth_file)
+
```
Comment (Schamper) [inline dissect/target/plugins/os/unix/log/auth.py:331]:
```suggestion
```

File: dissect/target/plugins/os/unix/log/auth.py [hunk 7]
```diff
@@ -1,60 +1,362 @@
...
+            else:
+                iterable = year_rollover_helper(auth_file, RE_TS, "%b %d %H:%M:%S", tzinfo)
+
+            for ts, line in iterable:
+                yield self._auth_log_builder.build_record(ts, auth_file, line)
+
+
+def iso_readlines(file: TargetPath) -> Iterator[tuple[datetime, str]]:
```
Comment (Schamper) [inline dissect/target/plugins/os/unix/log/auth.py:339]:
```suggestion
def iso_readlines(file: Path) -> Iterator[tuple[datetime, str]]:
```

File: dissect/target/plugins/os/unix/log/auth.py [hunk 8]
```diff
@@ -1,60 +1,362 @@
...
+
+            except ValueError as e:
+                log.warning("Unable to parse ISO timestamp in line: %s", line)
+                log.debug("", exc_info=e)
+                continue
+
+            yield ts, line
+
+
+def is_iso_fmt(file: TargetPath) -> bool:
```
Comment (Schamper) [inline dissect/target/plugins/os/unix/log/auth.py:360]:
```suggestion
def is_iso_fmt(file: Path) -> bool:
```

File: dissect/target/plugins/os/unix/log/auth.py [hunk 9]
```diff
@@ -241,15 +253,23 @@ def build_record(self, ts: datetime, source: Path, service: str, pid: int, messa
...
+            "pid": None,
+            "source": source,
+            "_target": self.target,
+        }
+
+        match = RE_LINE.match(line)
+        if match:
+            values = match.groupdict()
+            del values["ts"]
+            values["message"] = values["message"].strip()
```
Comment (Poeloe) [inline dissect/target/plugins/os/unix/log/auth.py:270]: Here I would suggest to do the stripping actions within the regex, such that this additional line is not needed. Something along the lines of `\s*?` before and after the `message` capturing group should work

File: dissect/target/plugins/os/unix/log/auth.py [hunk 10]
```diff
@@ -241,15 +253,23 @@ def build_record(self, ts: datetime, source: Path, service: str, pid: int, messa
...
+            "ts": ts,
+            "message": line,
+            "service": None,
+            "pid": None,
+            "source": source,
+            "_target": self.target,
+        }
+
+        match = RE_LINE.match(line)
+        if match:
```
Comment (Poeloe) [inline dissect/target/plugins/os/unix/log/auth.py:267]:
```suggestion
if match := RE_LINE.match(line)
```

File: dissect/target/plugins/os/unix/log/auth.py [hunk 11]
```diff
@@ -266,47 +286,76 @@ def _create_event_descriptor(self, record_fields) -> TargetRecordDescriptor:
...
+                yield self._auth_log_builder.build_record(ts, auth_file, line)
+
+
+def iso_readlines(file: TargetPath) -> Iterator[tuple[datetime, str]]:
+    """Iterator reading the provided auth log file in ISO format. Mimics ``year_rollover_helper`` behaviour."""
+
+    with open_decompress(file, "rt") as fh:
+        for line in fh:
+            if not (match := RE_TS_ISO.match(line)):
+                log.warning("No timestamp found in one of the lines in %s!", file)
```
Comment (Poeloe) [inline dissect/target/plugins/os/unix/log/auth.py:345]: Just a suggestion, but maybe you can pass the `Target` object and it's logger for more consistent logging output for the plugin

File: dissect/target/plugins/os/unix/log/auth.py [hunk 12]
```diff
@@ -161,7 +172,7 @@ class PkexecService(BaseService):
     """Class for parsing pkexec messages in the auth log"""
 
     PKEXEC_COMMAND_REGEX = re.compile(
-        r"(?P<user>.*?):\sExecuting\scommand\s"  # Starts with actual user -> user:
+        r"(?P<user>\S+?):\sExecuting\scommand\s"  # Starts with actual user -> user:
```
Comment (Poeloe) [inline dissect/target/plugins/os/unix/log/auth.py:170]: Out of curiosity, why is this changed to `\S+?`? Is it because there can't be any spaces in usernames or are there cases where there is a `:` in the username for this message?

File: dissect/target/plugins/os/unix/log/auth.py [hunk 13]
```diff
@@ -1,60 +1,362 @@
...
+
+log = logging.getLogger(__name__)
+
+_RE_TS = r"^[A-Za-z]{3}\s*[0-9]{1,2}\s[0-9]{1,2}:[0-9]{2}:[0-9]{2}"
+_RE_TS_ISO = r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{6}\+\d{2}:\d{2}"
+
+RE_TS = re.compile(_RE_TS)
+RE_TS_ISO = re.compile(_RE_TS_ISO)
+RE_LINE = re.compile(
+    rf"(?P<ts>{_RE_TS}|{_RE_TS_ISO})\s(?P<hostname>\S+)\s(?P<service>\S+?)(\[(?P<pid>\d+)\])?:\s(?P<message>.+)$"
```
Comment (Poeloe) [inline dissect/target/plugins/os/unix/log/auth.py:28]: The two suggestions corresponding to the suggestions further in the code:

Making a non-capturing group (and stripping spaces):
```suggestion
    rf"(?:{_RE_TS}|{_RE_TS_ISO})\s(?P<hostname>\S+)\s(?P<service>\S+?)(\[(?P<pid>\d+)\])?:\s+(?P<message>.+?)\s*$"
```
Or just check for the similar ending part of the timestamps and use `RE_LINE.search(line)` in the code below
```suggestion
    rf"\d{2}:\d{2}\s(?P<hostname>\S+)\s(?P<service>\S+?)(\[(?P<pid>\d+)\])?:\s+(?P<message>.+?)\s*$"
```

File: dissect/target/plugins/os/unix/log/auth.py [hunk 14]
```diff
@@ -241,15 +253,23 @@ def build_record(self, ts: datetime, source: Path, service: str, pid: int, messa
...
+            "service": None,
+            "pid": None,
+            "source": source,
+            "_target": self.target,
+        }
+
+        match = RE_LINE.match(line)
+        if match:
+            values = match.groupdict()
+            del values["ts"]
```
[the remark this answers is not in the document]
Comment (Poeloe) [reply dissect/target/plugins/os/unix/log/auth.py:269]: I understand, but that's for the purpose of matching the timestamp. If I'm correct, you don't need to capture the timestamp here, because you're deleting it anyway.

In that case, I would recommend to look at the two code suggestions I've done further down. In both cases, the timestamp is not "captured" but it is matched and therefore you don't need the del action here

File: dissect/target/plugins/os/unix/log/auth.py (hunk 13 above)
Comment (Schamper) [reply dissect/target/plugins/os/unix/log/auth.py:28]: This regex is getting long, can you use `re.VERBOSE` to split it up?
https://docs.python.org/3/library/re.html#re.VERBOSE

Comment (Poeloe) [reply dissect/target/plugins/os/unix/log/auth.py]: Used `re.VERBOSE` everywhere now

File: dissect/target/plugins/os/unix/log/auth.py [hunk 15]
```diff
@@ -1,62 +1,362 @@
...
+log = logging.getLogger(__name__)
+
+RE_TS = re.compile(r"^[A-Za-z]{3}\s*\d{1,2}\s\d{1,2}:\d{2}:\d{2}")
+RE_TS_ISO = re.compile(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{6}\+\d{2}:\d{2}")
+RE_LINE = re.compile(
+    r"""\d{2}:\d{2}\s                           # First match on the similar ending of the different timestamps
+        (?P<hostname>\S+)\s                     # The hostname
+        (?P<service>\S+?)(\[(?P<pid>\d+)\])?:   # The service with optionally the PID between brackets
+        \s*(?P<message>.+?)\s*$                 # The log message stripped from spaces left and right""",
+    re.X,
```
Comment (Schamper) [inline dissect/target/plugins/os/unix/log/auth.py:29]:
```suggestion
    r"""
    \d{2}:\d{2}\s                           # First match on the similar ending of the different timestamps
    (?P<hostname>\S+)\s                     # The hostname
    (?P<service>\S+?)(\[(?P<pid>\d+)\])?:   # The service with optionally the PID between brackets
    \s*(?P<message>.+?)\s*$                 # The log message stripped from spaces left and right
    """,
    re.VERBOSE,
```
Can you format them all to have the `r"""` and closing `"""` on separate lines? Also use `re.VERBOSE` everywhere instead of `re.X`.

**Review by Schamper — CHANGES_REQUESTED** [review]

File: dissect/target/plugins/os/unix/log/auth.py [hunk 16]
```diff
@@ -1,62 +1,374 @@
...
+            raise e
+
+    def build_record(self, ts: datetime, source: Path, line: str) -> TargetRecordDescriptor:
+        """Builds an ``AuthLog`` event record."""
+
+        record_fields = [
+            ("datetime", "ts"),
+            ("path", "source"),
+            ("string", "service"),
+            # PID should be string, since it can be "None"
```
Comment (Schamper) [inline dissect/target/plugins/os/unix/log/auth.py:270]: Comment is no longer relevant.

**Review by Schamper — CHANGES_REQUESTED** [review]

**Review by Schamper — APPROVED** [review]
